Update Nest integration to support Google Nest Device Access (new API)

allenporter opened the pull request:
<!--
  You are amazing! Thanks for contributing to our project!
  Please, DO NOT DELETE ANY TEXT from this template! (unless instructed).
-->
## Breaking change
<!--
  If your PR contains a breaking change for existing users, it is important
  to tell them what breaks, how to make it work again and why we did this.
  This piece of text is published with the release notes, so it helps if you
  write it towards our users, not us.
  Note: Remove this section if this PR is NOT a breaking change.
-->


## Proposed change
<!--
  Describe the big picture of your changes here to communicate to the
  maintainers why we should accept this pull request. If it fixes a bug
  or resolves a feature request, be sure to link to that issue in the
  additional information section.
-->

Google has released a new program to integrate with Nest called [Device Access](https://developers.google.com/nest/device-access) which uses the [Smart Device Management API](https://developers.google.com/nest/device-access/api).  The `nest` currently only supports the old API that is not accepting new users.   This pull request extends the existing Nest integration with support for the new API, using a CLOUD_PUSH style integration using a pubsub subscriber -- only supporting Temperature and Humidity sensors to reduce overall size of the request.


See https://github.com/home-assistant/core/pull/41119 for a previous pull request with a lot of discussion and design considerations, initially proposing a separate integration.

Assumptions:
- It preferred to have a single nest integration supporting both APIs
- Given there are two APIs, we're looking for simple ways to keep the two integrations as separate as possible
- Supporting both APIs at once is very complex.  As a result, we're not doing it for this initial integration, and unclear if this ever is worth it.  There is not currently a path forward for folks who want to use both the old and new APIs.

Other design considerations / components:
- A single config flow is used since a single flow for every integration is registered in global state.  Therefore, the config flow has some amount of complexity to deal with both APIs.  Otherwise, this new API integration with oauth would have only had a simple structure using the vanilla oauth2 config flow scaffolding with minimal extras like url params.
- Config and platform entry points branch off based on which API is used to separate files, to minimize intertwining the two code bases.
- This uses a new python library python-google-nest-sdm that supports polling or subscribing to a pubsub feed
- Every user must register their own developer sandbox and pay US$5.  See [Getting Started](https://developers.google.com/nest/device-access/get-started).
- Setup requires device registration, oauth credentials setup, enabling pubsub events, and creating a subscriber.  See [Subscribe to Events](https://developers.google.com/nest/device-access/subscribe-to-events) and [Google Cloud Platform: Pub/Sub: Subscriptions](https://console.cloud.google.com/cloudpubsub/subscriptions) to create a subscriber.
- Configuration is initially done in `configuration.yaml` then OAuth2 registration must be done by adding the integration in the UI
- The cloud push is implemented similar to the old `nest` API integration.  On startup, creates a GoogleNestSubscriber that has a pubsub subscriber underneath.  The subscriber keeps a DeviceManager up to date with the latest state, and also invokes a callback to notify HA of changes via a signal.  Each entity listens for the signal and refreshes itself to read the updated device properties.

## Type of change
<!--
  What type of change does your PR introduce to Home Assistant?
  NOTE: Please, check only 1! box!
  If your PR requires multiple boxes to be checked, you'll most likely need to
  split it into multiple PRs. This makes things easier and faster to code review.
-->

- [ ] Dependency upgrade
- [ ] Bugfix (non-breaking change which fixes an issue)
- [ ] New integration (thank you!)
- [X] New feature (which adds functionality to an existing integration)
- [ ] Breaking change (fix/feature causing existing functionality to break)
- [ ] Code quality improvements to existing code or addition of tests

## Example entry for `configuration.yaml`:
<!--
  Supplying a configuration snippet, makes it easier for a maintainer to test
  your PR. Furthermore, for new integrations, it gives an impression of how
  the configuration would look like.
  Note: Remove this section if this PR does not have an example entry.
-->

```yaml
# Example configuration.yaml
nest:
  project_id: your-project-id-abcd0349
  subscriber_id: projects/your-subscriber-id/subscriptions/enterprise-id-home-assistant-subscription
  client_id: your-client-id-abcdef.apps.googleusercontent.com
  client_secret: YOUR-CLIENT-SECRET
```
Note that many configuration options for the old API are currently ignored such as filtering, as well as multiple unsupported services.  Some options will be supported in future versions, while others will not be supported due to differences in API support.

## Additional information
<!--
  Details are important, and help maintainers processing your PR.
  Please be sure to fill out additional details, if applicable.
-->

- This PR fixes or closes issue: fixes #
- This PR is related to issue: 
- Link to documentation pull request: https://github.com/home-assistant/home-assistant.io/pull/14855

## Checklist
<!--
  Put an `x` in the boxes that apply. You can also fill these out after
  creating the PR. If you're unsure about any of them, don't hesitate to ask.
  We're here to help! This is simply a reminder of what we are going to look
  for before merging your code.
-->

- [X] The code change is tested and works locally.
- [X] Local tests pass. **Your PR cannot be merged unless tests pass**
- [X] There is no commented out code in this PR.
- [X] I have followed the [development checklist][dev-checklist]
- [X] The code has been formatted using Black (`black --fast homeassistant tests`)
- [x] Tests have been added to verify that the new code works.

If user exposed functionality or configuration variables are added/changed:

- [x] Documentation added/updated for [www.home-assistant.io][docs-repository]

If the code communicates with devices, web services, or third-party tools:

- [x] The [manifest file][manifest-docs] has all fields filled out correctly.  
      Updated and included derived files by running: `python3 -m script.hassfest`.
- [x] New or updated dependencies have been added to `requirements_all.txt`.  
      Updated by running `python3 -m script.gen_requirements_all`.
- [x] Untested files have been added to `.coveragerc`.

The integration reached or maintains the following [Integration Quality Scale][quality-scale]:
<!--
  The Integration Quality Scale scores an integration on the code quality
  and user experience. Each level of the quality scale consists of a list
  of requirements. We highly recommend getting your integration scored!
-->

- [X] No score or internal
- [ ] 🥈 Silver
- [ ] 🥇 Gold
- [ ] 🏆 Platinum

<!--
  This project is very active and we have a high turnover of pull requests.

  Unfortunately, the number of incoming pull requests is higher than what our
  reviewers can review and merge so there is a long backlog of pull requests
  waiting for review. You can help here!
  
  By reviewing another pull request, you will help raise the code quality of
  that pull request and the final review will be faster. This way the general
  pace of pull request reviews will go up and your wait time will go down.
  
  When picking a pull request to review, try to choose one that hasn't yet
  been reviewed.

  Thanks for helping out!
-->

To help with the load of incoming pull requests:

- [X] I have reviewed two other [open pull requests][prs] in this repository.

[prs]: https://github.com/home-assistant/core/pulls?q=is%3Aopen+is%3Apr+-author%3A%40me+-draft%3Atrue+-label%3Awaiting-for-upstream+sort%3Acreated-desc+review%3Anone

<!--
  Thank you for contributing <3

  Below, some useful links you could explore:
-->
[dev-checklist]: https://developers.home-assistant.io/docs/en/development_checklist.html
[manifest-docs]: https://developers.home-assistant.io/docs/en/creating_integration_manifest.html
[quality-scale]: https://developers.home-assistant.io/docs/en/next/integration_quality_scale_index.html
[docs-repository]: https://github.com/home-assistant/home-assistant.io

allenporter committed:
Update nest integration with new API auth flow

Add Google Nest Device Access program authentication flow using OAuth.  This
supports both APIs since the new API does not have all the features of the
old API.  This adds a sensor that polls as a placeholder that doesn't work
yet end-to-end, which will be replaced with a cloud push model.

See https://developers.google.com/nest/device-access/registration for details
on the API.

allenporter committed:
Update nest integration with new API auth flow

Add Google Nest Device Access program authentication flow using OAuth.  This
supports both APIs since the new API does not have all the features of the
old API.  This adds a sensor that polls as a placeholder that doesn't work
yet end-to-end, which will be replaced with a cloud push model.

See https://developers.google.com/nest/device-access/registration for details
on the API.

allenporter committed:
Update Nest Device Access to use push events

Use the event based subscriber in the python-google-nest-sdm integration
to update sensor values vai CLOUD_PUSH rather than polling.

The overall strategy is to create a GoogleNestSubscriber which keeps state
updated in a DeviceManager and invokes a callback when events have been
received.  Each device created registers with a signal handler to refresh its
state, and the subscriber callback invokes that signal handler.

allenporter committed:
Update Nest Device Access to use push events

Use the event based subscriber in the python-google-nest-sdm integration
to update sensor values vai CLOUD_PUSH rather than polling.

The overall strategy is to create a GoogleNestSubscriber which keeps state
updated in a DeviceManager and invokes a callback when events have been
received.  Each device created registers with a signal handler to refresh its
state, and the subscriber callback invokes that signal handler.

allenporter committed:
Remove verbose logging

allenporter committed:
Update documentation and comment style.

Update comments, whitespace, and documentation to be more consistent with
the rest of the surrounding code.

allenporter committed:
Upgrade to latest python-nest-sdm resolving pylint

Switch to the latest version, fixing bug found by python3.7 lint.

allenporter committed:
Remove incorrect calls to super() methods for SDM

Remove dead code inside of the legacy flows where no equivalents
exist for the new API.  Still leaving the SDM API check just in
case.  Found by py37 lint errors on CI.

allenporter committed:
Simplify debug messages to resolve py37 lint errors

allenporter committed:
Update to google-nest-python-sdm 0.1.6

Use latest version that results some bugs in device manager initialization.

poudenes commented:
Will this new integration support also the Nest Protect fully?

SeraphimSerapis commented:
@poudenes Nest Protect is currently not covered by the Nest SDM API (you can see the supported devices [here](https://developers.google.com/nest/device-access/api)).

vincebalsamo commented:
Looks like this is only supporting thermostats to start. Any plans to add cameras?

allenporter commented:
Yes, that will happen.  My impression is that folks prefer smaller, focused, pull requests, hence just supporting one platform for the first attempt.  The underlying python library already has support for camera events.

tobyworks commented:
Can i manually grab this PR and put it in my own home assistant? without waiting for it to get merged ?

allenporter committed:
Add tests for the thermostat sensors

Test uses a fake susbcriber with a fake device manager that simluates
subscribing to the pubsub feed and receiving device updates.

allenporter committed:
Update Nest sensor for isort, after branch merge.

allenporter committed:
Remove Nest SDM sensor debug logging for pylint37

allenporter commented:
Since the initial pull request, I have also added unit tests for the new sensors, and have updated the [documentation PR](https://github.com/home-assistant/home-assistant.io/pull/14855) in order to round out this request to make it even more ready for review.

bdraco commented:
Thank you for your contribution thus far! 🎖 Since this is a significant contribution, we would appreciate you'd added yourself to the list of code owners for this integration. ❤️ 

Please, add your GitHub username to the `manifest.json` of this integration.

For more information about "code owners", see: [Architecture Decision Record 0008: Code owners](https://github.com/home-assistant/architecture/blob/master/adr/0008-code-owners.md).

balloob commented on homeassistant/components/nest/__init__.py:
```diff
@@ -94,10 +120,128 @@
+        auth, config[CONF_PROJECT_ID], config[CONF_SUBSCRIBER_ID]
+    )
+    subscriber.set_update_callback(SignalUpdateCallback(hass))
+    hass.async_create_task(subscriber.start_async())
```
If this is a long living task, you should use `hass.loop.create_task`. `hass.async_create_task` used during startup of Home Assistant will keep track of the task and will block startup from being finished.
```suggestion
    hass.loop.create_task(subscriber.start_async())
```

balloob replied:
`SignalUpdateCallback.handle_event` calls `dispatcher_send`, which is a synchronous function. Here I see that we're running the subscriber async. Does the subscriber trigger handle events in sync context?

allenporter replied:
This subscriber is a wrapper around the Google subscriber Client subscribe method:
https://googleapis.dev/python/pubsub/latest/subscriber/api/client.html#google.cloud.pubsub_v1.subscriber.client.Client.subscribe
Note: we are not currently using a scheduler.

So the subscriber is started async which runs a background thread.   It then invokes our callback, which:
(1) updates the state in the DeviceManager and (2) invokes our handle_event  then (3) once that is done, our subscriber calls ack() on the message.

I think we'd be OK with blocking the subscriber thread, and waiting for the response before acking the message.

I copied this dispatcher_send with a signal from the existing nest integration, which has a somewhat similar model.

Here are some options I can see:
1) Might there be a simple way to avoid using signal at all? Each Device could have its own "add_update_callback" that gets invoked directly when the DeviceManager updates its traits or sends it an event (e.g. person_motion, etc).  I think this might be needed anyway once we get to implementing cameras or binary_sensor, but i'm not certain.
2) Switch this all to be async
3) I think we may be OK with blocking the subscriber thread, and ACKing once this returns.  Maybe extra comments explaining the sync vs async model of the whole thing.

Open to other ideas if you have them, or where you'd like to see extra comments.

balloob commented on homeassistant/components/nest/sensor_sdm.py:
```diff
@@ -0,0 +1,169 @@
+    def device_info(self):
+        """Return device specific attributes."""
+        return {
+            "identifiers": {(DOMAIN, self._device.name)},
```
We shouldn't use a name as an identifier. Do we have access to an ID or a serial number?

allenporter replied:
This actually is a unique device identifier, however, the API calls it a name.
See an example of the output here: https://developers.google.com/nest/device-access/api#get

I kept the field called "name" in the python API following the guidance in https://developers.home-assistant.io/docs/api_lib_data_models/

This same self._device.name field is used above in the unique_id property and it has a clarifying comment there.  To address your comment, I also added the same comment here, to clarify.

balloob replied:
Ok!

balloob commented on homeassistant/components/nest/sensor_sdm.py:
```diff
@@ -0,0 +1,169 @@
+    async def async_added_to_hass(self):
+        """Run when entity is added to register update signal handler."""
+
+        async def async_update_state():
+            """Update sensor state."""
+            await self.async_update_ha_state(True)
+
+        # Event messages trigger the SIGNAL_NEST_UPDATE, which is intercepted
+        # here to re-fresh the signals from _device.  Unregister this callback
+        # when the entity is removed.
+        self.async_on_remove(
+            async_dispatcher_connect(self.hass, SIGNAL_NEST_UPDATE, async_update_state)
+        )
```
This will call `await self.async_update()` and then `self.async_write_ha_state()`. Is there state to be updated or can we just run `self.async_write_ha_state()` directly inside a callback? 
```suggestion
        # Event messages trigger the SIGNAL_NEST_UPDATE, which is intercepted
        # here to re-fresh the signals from _device.  Unregister this callback
        # when the entity is removed.
        self.async_on_remove(
            async_dispatcher_connect(self.hass, SIGNAL_NEST_UPDATE, self.async_write_ha_state)
        )
```

allenporter replied:
The _device itself should be updated by the time we get here.  This looks good to me.

allenporter replied:
I tried this out and it ended up failing tests.  I'm not totally sure I understand what could be improved here.

By the time this runs, _device has been updated with the latest state by the DeviceManager in the pubsub callback.  Is this a threadsafety issue here? Python threading model isn't totally my specialty.

balloob commented:
Already looking really good. Just some minor comments.

allenporter committed:
Add myself as a nest code owner

Following guidance in https://github.com/home-assistant/architecture/blob/master/adr/0008-code-owners.md
as recommended by bdraco

allenporter committed:
Add clarifying comment for nest sensor device id

This device unique idenifier is returned under the API field name,
so add a comment clarifying what it actually is.

allenporter committed:
Use hass.loop.create_task to start nest subscriber

allenporter committed:
Simplify nest event updates

Invoke async_update_state directly from callback since a force refresh is not needed given all device state is already udpated.

Co-authored-by: Paulus Schoutsen <paulus@home-assistant.io>

allenporter committed:
Apply black reformatting on nest sensor

allenporter committed:
Revert change to nest sensor update state

Applied changes over review, but tests ended up failing.  Reverting.

balloob reviewed (approved):
Looks good! Let's fix the async_write_ha_state later so this can go in the beta

mrdarrengriffin commented:
Wooo! Nice work guys!

MartinHjelmare commented on homeassistant/components/nest/__init__.py:
```diff
@@ -94,10 +120,128 @@
+        auth, config[CONF_PROJECT_ID], config[CONF_SUBSCRIBER_ID]
+    )
+    subscriber.set_update_callback(SignalUpdateCallback(hass))
+    hass.loop.create_task(subscriber.start_async())
```
We can use `asyncio.create_task` with Python 3.7+.

allenporter replied:
Addressing in PR 42250

MartinHjelmare commented on homeassistant/components/nest/config_flow.py:
```diff
@@ -47,22 +62,57 @@ class CodeInvalid(NestAuthError):
     async def async_step_init(self, user_input=None):
         """Handle a flow start."""
+        if self.is_sdm_api():
+            return None
```
This would raise a TypeError in the flow manager. Is that intended?

https://github.com/home-assistant/core/blob/c677489535be5d44efaf6baa01a96f2dddd4a1af/homeassistant/data_entry_flow.py#L214

MartinHjelmare commented on homeassistant/components/nest/sensor.py:
```diff
@@ -1,208 +1,17 @@
+) -> None:
+    """Set up the sensors."""
+    if DATA_SDM not in entry.data:
+        return await async_setup_legacy_entry(hass, entry, async_add_entities)
```
Nothing is checking the return value of this function.

MartinHjelmare commented on homeassistant/components/nest/sensor_sdm.py:
```diff
@@ -0,0 +1,170 @@
+        self._device = device
+
+    @property
+    def should_pool(self) -> bool:
```
Typo -> `should_poll`

MartinHjelmare commented on homeassistant/components/nest/sensor_sdm.py:
```diff
@@ -0,0 +1,170 @@
+
+        async def async_update_state():
+            """Update sensor state."""
+            await self.async_update_ha_state(True)
```
We don't define an `update` or `async_update` method so there's no point in passing `True`.

allenporter replied:
I mistakenly thought tests failed when inlining the async_update_ha_state, but tried it again and it works.  Addressed in PR Addressing in PR 42250

MartinHjelmare reviewed (commented):
Please address the comments in a new PR. Thanks!
